### PR TITLE
add support for Flysystem V2

### PR DESCRIPTION
**Pull request by Warxcell**
| Q | A
| --- | ---
| Branch? | 2
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Tests pass? | yes/no
| Fixed tickets | #1308
| License | MIT
| Doc PR | 

Add support for https://flysystem.thephpleague.com/v2/docs/what-is-new/

### CODE REVIEW

**Comment by mynameisbogdan** [comment]
Hello @Warxcell,

Can you please take a look at FlysystemResolver as well?

Thank you.

**Comment by lsmith77** [comment]
there seems to be some coding style issues that need to be addressed

**Comment by Warxcell** [comment]
Yes, but it doesn't tell anything useful. For example diff and files.
```
Run php-cs-fixer fix --dry-run --format checkstyle | cs2pr
  php-cs-fixer fix --dry-run --format checkstyle | cs2pr
  shell: /usr/bin/bash -e {0}
Loaded config default from "/home/runner/work/LiipImagineBundle/LiipImagineBundle/.php_cs.dist".
Warning: Found violation(s) of type: phpdoc_order
Warning: Found violation(s) of type: phpdoc_separation
Warning: Found violation(s) of type: phpdoc_trim
Warning: Found violation(s) of type: phpdoc_align
Error: Process completed with exit code 1.
```

**Comment by lsmith77** [comment]
ah .. I see .. @fbourigault I guess we need to adjust this action to give useful output.

**Comment by fbourigault** [comment]
Those should appear in this MR “Files changed” tab.

**Comment by Warxcell** [comment]
> Those should appear in this MR “Files changed” tab.

Not quite. I will see them tomorrow locally.

**Comment by lsmith77** [comment]
@fbourigault it just shows this

grafik

not sure if we had used https://styleci.io/ on this repo in the past and if we removed it when we added GithubActions but it shows the specific issues and even makes a PR if you want.

**Comment by fbourigault** [comment]
My bad PHP-CS-Fixer doesn't report which line contains the error.

I opened https://github.com/liip/LiipImagineBundle/pull/1355 to remove use of cs2pr and display diff in action output.

**Comment by lsmith77** [comment]
thx .. merged to 2.x .. can you rebase @Warxcell ?

**Comment by Warxcell** [comment]
Done. BTW can you give me a hand with tests - we need to run test with version 1 and version 2 installed. How can we achieve that?

File: composer.json [hunk 1]
```diff
@@ -44,7 +44,8 @@
         "symfony/phpunit-bridge": "^5.2",
         "symfony/templating": "^3.4|^4.3|^5.0",
         "symfony/validator": "^3.4|^4.3|^5.0",
-        "symfony/yaml": "^3.4|^4.3|^5.0"
+        "symfony/yaml": "^3.4|^4.3|^5.0",
+        "friendsofphp/php-cs-fixer": "*"
```
Comment (fbourigault) [inline composer.json:48]: You have to remove this line as we don't use PHP-CS-Fixer as a dev dependency anymore.

**Comment by fbourigault** [comment]
You have to find something that make a difference between version 1 and 2 of Flysystem  (V1 has `League\Flysystem\FilesystemInterface` interface but V2 hasn't)and then skip V1 tests when V2 is detected and skip V2 tests when V1 is detected.

**Comment by Warxcell** [comment]
> You have to find something that make a difference between version 1 and 2 of Flysystem (V1 has `League\Flysystem\FilesystemInterface` interface but V2 hasn't)and then skip V1 tests when V2 is detected and skip V2 tests when V1 is detected.

Already did that, but now V1 tests fails. Or they will be skipped and never ran.

File: composer.json (hunk 1 above)
Comment (Warxcell) [reply composer.json:48]: But I don't have global php-cs-fixer and the only way to use it is as dependency. That way package is fully packaged with all dependencies. Isn't it better?

**Comment by Warxcell** [comment]
I think travis must be configured to run tests on both versions.

**Comment by lsmith77** [comment]
we are using GithubActions .. but yeah I think we likely have to have a separate build for this.
I propose we test flysystem v2 by default and have one separate build where we test flysystem v1

wdyt @fbourigault ?

**Comment by fbourigault** [comment]
You have to skip V1 tests when the interface is not installed. There is 0 chance to successfully run Flysystem V1 tests with highest dependencies!

**Comment by Warxcell** [comment]
image

Old tests are not working.

**Comment by fbourigault** [comment]
Have you tried to skip FlysystemV1 related tests when V2 is installed?

Something like this in `FlysystemLoaderTest` and `FlysystemResolverTest` should do the work:
```php
        if (!class_exists(FilesystemInterface::class)) {
            $this->markTestSkipped('Requires the league/flysystem:^1.0 package.');
        }
```

**Comment by Warxcell** [comment]
Done, skipped tests for V1, but why Coveralls complains about old tests now?

**Comment by lsmith77** [comment]
I guess the reduction in coverage is a side effect of now using either V1 or V2.
I retriggered scrutinizer

**Comment by fbourigault** [comment]
There is something wrong. If you look at https://github.com/liip/LiipImagineBundle/pull/1349/checks?check_run_id=1914419168, both V1 and V2 tests are skipped.

**Comment by Warxcell** [comment]
Fixed that. Seems `class_exists` doesn't work for interfaces. I always thought it works. 

Now overalls is good but some of the tests fails. Will review them later today.

**Comment by Warxcell** [comment]
Finally, all green :)

File: Tests/Binary/Loader/FlysystemV2LoaderTest.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+use League\Flysystem\FilesystemOperator;
+use League\Flysystem\Local\LocalFilesystemAdapter;
+use Liip\ImagineBundle\Binary\Loader\FlysystemV2Loader;
+use Liip\ImagineBundle\Binary\Loader\LoaderInterface;
+use Liip\ImagineBundle\Exception\Binary\Loader\NotLoadableException;
+use Liip\ImagineBundle\Tests\AbstractTest;
+use Symfony\Component\Mime\MimeTypes;
+
+/**
+ * @requires PHP 7.1
```
Comment (mynameisbogdan) [inline Tests/Binary/Loader/FlysystemV2LoaderTest.php]: [league/flysystem v2 requires PHP 7.2](https://github.com/thephpleague/flysystem/blob/7cbbb7222e8d8a34e71273d243dfcf383ed6779f/composer.json#L20)

```suggestion
 * @requires PHP 7.2
```

File: composer.json (hunk 1 above)
Comment (mynameisbogdan) [reply composer.json:48]: Why don't you use this instead `composer global require friendsofphp/php-cs-fixer`, export PATH="$PATH:$HOME/.composer/vendor/bin" and `php-cs-fixer`?

**Comment by lsmith77** [comment]
LGTM .. @fbourigault WDYT?

**Comment by fbourigault** [comment]
Could you remove `friendsofphp/php-cs-fixer` from `require-dev` now everything is ok?

Once removed, this one could be merged.

**Comment by mynameisbogdan** [comment]
I'm sorry to ask, but isn't #1357 a more proper implementation?

**Comment by fbourigault** [comment]
Why #1357 would be better?

**Comment by mynameisbogdan** [comment]
@fbourigault because it handles the integration for the loader and resolver by adding service prototypes for them, and the configuration factories detects if you're using v1 or v2 to know which service to inject.

**Comment by lsmith77** [comment]
Thank you @Warxcell for working through this process with us! You helped mature our new Github Actions based setup and also bring Flysystem2 support to the Bundle.

@mynameisbogdan also thank you for your PR. I will merge this PR here and would then appreciate if you could rebase your PR to add the additional pieces on top of @Warxcell’s work.

**Pull request merged**